File: android/app/src/main/java/com/openburnbar/ui/settings/SettingsRouter.kt
package com.openburnbar.ui.settings

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue

/**
 * Drives Settings search navigation on Android. Single source of truth for
 * the search query, current sub-screen, pending anchor, pending focus, and
 * highlight target.
 *
 * Lifecycle: scoped to the `SettingsRoot` composable via `remember`.
 */
class SettingsRouter {
    /** Free-form search query. */
    var query: String by mutableStateOf("")

    /** Current sub-screen the Settings tree is showing. */
    var page: SettingsPageRoute by mutableStateOf(SettingsPageRoute.ROOT)

    /** Anchor id the next composable should scroll to on appear. */
    var pendingAnchor: String? by mutableStateOf(null)

    /** Focus id the next composable should hand to its FocusRequester. */
    var pendingFocus: String? by mutableStateOf(null)

    /** Anchor that should paint a brief halo on arrival. */
    var highlightedAnchor: String? by mutableStateOf(null)

    val isSearching: Boolean
        get() = query.isNotBlank()

    /** Drive navigation to a search result row. */
    fun navigate(item: SettingsItem) {
        pendingAnchor = item.anchorId
        highlightedAnchor = item.anchorId
        pendingFocus = item.focusId
        page = item.pageRoute
        query = ""
    }

    /** Clear all transient routing state. */
    fun reset() {
        query = ""
        pendingAnchor = null
        pendingFocus = null
        highlightedAnchor = null
        page = SettingsPageRoute.ROOT
    }

    /** Destination calls this after it scrolls to the anchor. */
    fun consumePendingAnchor(anchor: String) {
        if (pendingAnchor == anchor) pendingAnchor = null
    }

    /** Destination calls this after it latches the focus state. */
    fun consumePendingFocus(focus: String) {
        if (pendingFocus == focus) pendingFocus = null
    }

    /** Destinations call this once the highlight animation completes. */
    fun clearHighlight(anchor: String) {
        if (highlightedAnchor == anchor) highlightedAnchor = null
    }
}

File: OpenBurnBarMobile/Settings/Search/SettingsRouter.swift
import SwiftUI

// MARK: - Settings Router (iOS)

/// Drives programmatic navigation inside the iOS Settings hub when the
/// `.searchable` field is active.
///
/// Flow:
/// 1. User types into the Form's `.searchable` field — `query` updates.
/// 2. When `query` is non-empty, the hub renders `SettingsSearchResultsView`
///    in place of the Form (or above it depending on placement).
/// 3. User taps a row → `navigate(to:)` pushes the page route onto `path`
///    and stashes `pendingAnchor` / `pendingFocus`.
/// 4. The destination view consumes the pending anchor with a
///    `ScrollViewReader` proxy and focuses the bound `@FocusState` if a
///    focus id was supplied.
@Observable
@MainActor
final class SettingsRouter {

    /// Free-form search query.
    var query: String = ""

    /// `NavigationStack` path the hub binds to.
    var path: [SettingsPageRoute] = []

    /// Anchor id destinations should scroll to on appear. Cleared by
    /// `consumePendingAnchor(_:)`.
    var pendingAnchor: String?

    /// Focus id destinations should latch onto. Cleared by
    /// `consumePendingFocus(_:)`.
    var pendingFocus: String?

    /// Anchor id whose row should pulse-highlight while it animates into
    /// view. Cleared automatically after ~1.4s.
    var highlightedAnchor: String?

    var isSearching: Bool {
        !query.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
    }

    func results(in items: [SettingsItem] = SettingsManifest.all) -> [SettingsItem] {
        SettingsSearchEngine.search(query, in: items)
    }

    /// Push the destination view and prime the anchor/focus for it.
    func navigate(to item: SettingsItem) {
        pendingAnchor = item.anchorID
        highlightedAnchor = item.anchorID
        pendingFocus = item.focusID

        switch item.pageRoute {
        case .hubRoot:
            // Row already lives on the root form — make sure no sub-page is
            // covering it.
            path.removeAll()
        case .cloud, .providerConnections, .hermes, .pi, .chatTiles, .media:
            path = [item.pageRoute]
        }

        // Clear the query so the hub shows the destination (not the
        // results list) once routing settles.
        query = ""

        scheduleHighlightClear(for: item.anchorID, after: 1.4)
    }

    /// Reset state without selecting anything (e.g. user dismissed search).
    func reset() {
        query = ""
        pendingAnchor = nil
        pendingFocus = nil
        highlightedAnchor = nil
        path.removeAll()
    }

    func consumePendingAnchor(_ anchor: String) {
        if pendingAnchor == anchor { pendingAnchor = nil }
    }

    func consumePendingFocus(_ focus: String) {
        if pendingFocus == focus { pendingFocus = nil }
    }

    private func scheduleHighlightClear(for anchor: String, after seconds: TimeInterval) {
        let target = anchor
        Task { @MainActor [weak self] in
            try? await Task.sleep(nanoseconds: UInt64(seconds * 1_000_000_000))
            guard let self else { return }
            if self.highlightedAnchor == target { self.highlightedAnchor = nil }
        }
    }
}
